src/compiler/ng-jest-compiler.spec.ts

```diff
@@ -1,34 +1,587 @@
+                ];
+                exports.MyDir = MyDir = tslib_1.__decorate([
+                    (0, core_1.Directive)()
+                ], MyDir);
+            `);
+        });
+
+        it.failing('should not capture constructor parameter types when not resolving to a value', () => {
+            // This test doesn't pass because `transpileModule` can't resolve `import *`
+            const { code, diagnostics } = transformCjs(`
+                import {Directive, Inject} from '@angular/core';
+                import * as angular from './external';
+                import {IOverlay, KeyCodes} from './external';
+                import TypeFromDefaultImport from './external';
+
+                @Directive()
+                export class MyDir {
+                    constructor(@Inject('$state') param: angular.IState,
+                                @Inject('$overlay') other: IOverlay,
+                                @Inject('$default') fromDefaultImport: TypeFromDefaultImport,
+                                @Inject('$keyCodes') keyCodes: KeyCodes) {}
+                }
+            `);
+
+            expect(diagnostics.length).toBe(0);
+            expect(code).not.toContain('external');
+            expect(code).toContain(dedent`
+                MyDir.ctorParameters = () => [
+                    { type: undefined, decorators: [{ type: core_1.Inject, args: ['$state',] }] },
+                    { type: undefined, decorators: [{ type: core_1.Inject, args: ['$overlay',] }] },
+                    { type: undefined, decorators: [{ type: core_1.Inject, args: ['$default',] }] },
+                    { type: undefined, decorators: [{ type: core_1.Inject, args: ['$keyCodes',] }] }
+                ];
+                exports.MyDir = MyDir = tslib_1.__decorate([
+                    (0, core_1.Directive)()
+                ], MyDir);
+            `);
+        });
+
+        it('should capture a non-const enum used as a constructor type', () => {
+            const { code, diagnostics } = transformCjs(`
+                import {Component} from '@angular/core';
+
+                export enum Values {A, B};
+
+                @Component({template: 'hello'})
+                export class MyComp {
+                    constructor(v: Values) {}
+                }
+            `);
+
+            expect(diagnostics.length).toBe(0);
+            expect(code).toContain(dedent`
+                MyComp.ctorParameters = () => [
+                    { type: Values }
+                ];
+                exports.MyComp = MyComp = tslib_1.__decorate([
+                    (0, core_1.Component)({ template: 'hello' })
+                ], MyComp);
+            `);
+        });
+
+        it.failing(
+            'should allow for type-only references to be removed with `emitDecoratorMetadata` from custom decorators',
+            () => {
+                // This test only passes when using `import type { ExternalInterface } from './external-interface';`
```

Comment:
This one makes sense to me because with `isolatedModules` TS shouldn't know that it's an interface and not eligible to be moved into a value reference position.

Reply:
`typeof external_interface_1.ExternalInterface` sounds wrong to me. Interfaces are erased by TypeScript, so this expression can't work at runtime. I would expect this to be something like `Object` in this scenario since it can't be any more specific about the type.

I don't think Angular uses `design:returntype` anywhere. I think we only care about constructor parameters. So it probably doesn't matter what's actually in the `__metadata` call, as long as it doesn't throw.